In [1]:
#Importing required libraries
import pandas as pd
import numpy as np

In this notebook I am gonna apply Conteny based filtering to the famous MovieLens 20M dataset and recommend similar movies to a

movie which the user is watching

Lets start by exploring our datasets

In [2]:
ratings = pd.read_csv('rating.csv')
ratings

userId  movieId  rating            timestamp
0              1        2     3.5  2005-04-02 23:53:47
1              1       29     3.5  2005-04-02 23:31:16
2              1       32     3.5  2005-04-02 23:33:39
3              1       47     3.5  2005-04-02 23:32:07
4              1       50     3.5  2005-04-02 23:29:40
...          ...      ...     ...                  ...
20000258  138493    68954     4.5  2009-11-13 15:42:00
20000259  138493    69526     4.5  2009-12-03 18:31:48
20000260  138493    69644     3.0  2009-12-07 18:10:57
20000261  138493    70286     5.0  2009-11-13 15:42:24
20000262  138493    71619     2.5  2009-10-17 20:25:36

[20000263 rows x 4 columns]

In [3]:
movies = pd.read_csv('movie.csv')
movies

movieId                               title  \
0            1                    Toy Story (1995)   
1            2                      Jumanji (1995)   
2            3             Grumpier Old Men (1995)   
3            4            Waiting to Exhale (1995)   
4            5  Father of the Bride Part II (1995)   
...        ...                                 ...   
27273   131254        Kein Bund für's Leben (2007)   
27274   131256       Feuer, Eis & Dosenbier (2002)   
27275   131258                  The Pirates (2014)   
27276   131260                 Rentun Ruusu (2001)   
27277   131262                    Innocence (2014)   

                                            genres  
0      Adventure|Animation|Children|Comedy|Fantasy  
1                       Adventure|Children|Fantasy  
2                                   Comedy|Romance  
3                             Comedy|Drama|Romance  
4                                           Comedy  
...                                            ...  
27273                                       Comedy  
27274                                       Comedy  
27275                                    Adventure  
27276                           (no genres listed)  
27277                     Adventure|Fantasy|Horror  

[27278 rows x 3 columns]

In [4]:
ratings = pd.merge(movies,ratings,on ="movieId",how ="inner") #Merging ratings dataframe and movies dataframe so as to get relevant information
ratings['index'] = range(1, len(ratings) + 1)
ratings

movieId                          title  \
0               1               Toy Story (1995)   
1               1               Toy Story (1995)   
2               1               Toy Story (1995)   
3               1               Toy Story (1995)   
4               1               Toy Story (1995)   
...           ...                            ...   
20000258   131254   Kein Bund für's Leben (2007)   
20000259   131256  Feuer, Eis & Dosenbier (2002)   
20000260   131258             The Pirates (2014)   
20000261   131260            Rentun Ruusu (2001)   
20000262   131262               Innocence (2014)   

                                               genres  userId  rating  \
0         Adventure|Animation|Children|Comedy|Fantasy       3     4.0   
1         Adventure|Animation|Children|Comedy|Fantasy       6     5.0   
2         Adventure|Animation|Children|Comedy|Fantasy       8     4.0   
3         Adventure|Animation|Children|Comedy|Fantasy      10     4.0   
4         Adventure|Animation|Children|Comedy|Fantasy      11     4.5   
...                                               ...     ...     ...   
20000258                                       Comedy   79570     4.0   
20000259                                       Comedy   79570     4.0   
20000260                                    Adventure   28906     2.5   
20000261                           (no genres listed)   65409     3.0   
20000262                     Adventure|Fantasy|Horror  133047     4.0   

                    timestamp     index  
0         1999-12-11 13:36:47         1  
1         1997-03-13 17:50:52         2  
2         1996-06-05 13:37:51         3  
3         1999-11-25 02:44:47         4  
4         2009-01-02 01:13:41         5  
...                       ...       ...  
20000258  2015-03-30 19:32:59  20000259  
20000259  2015-03-30 19:48:08  20000260  
20000260  2015-03-30 19:56:32  20000261  
20000261  2015-03-30 19:57:46  20000262  
20000262  2015-03-30 20:39:26  20000263  

[20000263 rows x 7 columns]

In [5]:
ratings.drop_duplicates(subset = ['title','movieId'], inplace=True)
ratings

movieId                               title  \
0               1                    Toy Story (1995)   
49695           2                      Jumanji (1995)   
71938           3             Grumpier Old Men (1995)   
84673           4            Waiting to Exhale (1995)   
87429           5  Father of the Bride Part II (1995)   
...           ...                                 ...   
20000258   131254        Kein Bund für's Leben (2007)   
20000259   131256       Feuer, Eis & Dosenbier (2002)   
20000260   131258                  The Pirates (2014)   
20000261   131260                 Rentun Ruusu (2001)   
20000262   131262                    Innocence (2014)   

                                               genres  userId  rating  \
0         Adventure|Animation|Children|Comedy|Fantasy       3     4.0   
49695                      Adventure|Children|Fantasy       1     3.5   
71938                                  Comedy|Romance       2     4.0   
84673                            Comedy|Drama|Romance      41     2.0   
87429                                          Comedy      12     2.0   
...                                               ...     ...     ...   
20000258                                       Comedy   79570     4.0   
20000259                                       Comedy   79570     4.0   
20000260                                    Adventure   28906     2.5   
20000261                           (no genres listed)   65409     3.0   
20000262                     Adventure|Fantasy|Horror  133047     4.0   

                    timestamp     index  
0         1999-12-11 13:36:47         1  
49695     2005-04-02 23:53:47     49696  
71938     2000-11-21 15:34:49     71939  
84673     2000-02-27 23:29:58     84674  
87429     1997-03-22 20:49:34     87430  
...                       ...       ...  
20000258  2015-03-30 19:32:59  20000259  
20000259  2015-03-30 19:48:08  20000260  
20000260  2015-03-30 19:56:32  20000261  
20000261  2015-03-30 19:57:46  20000262  
20000262  2015-03-30 20:39:26  20000263  

[26744 rows x 7 columns]

In [6]:
tags = pd.read_csv('tag.csv')
tags.drop('timestamp',axis=1)

userId  movieId            tag
0           18     4141    Mark Waters
1           65      208      dark hero
2           65      353      dark hero
3           65      521  noir thriller
4           65      592      dark hero
...        ...      ...            ...
465559  138446    55999        dragged
465560  138446    55999  Jason Bateman
465561  138446    55999         quirky
465562  138446    55999            sad
465563  138472      923  rise to power

[465564 rows x 3 columns]

In [7]:
genome_score = pd.read_csv('genome_scores.csv')

In [8]:
genome_tag = pd.read_csv('genome_tags.csv')

We have loaded all our datasets from the folder. 

The first file has following features:  userid,movieid,rating,timestamp

second file has movieid,title,genres

third file has userid,movieid,tag,timestamp

fourth file has movieid,tagid,relevance

fifth file has tag id and tag 


In [9]:
ratings.head(5)

movieId                               title  \
0            1                    Toy Story (1995)   
49695        2                      Jumanji (1995)   
71938        3             Grumpier Old Men (1995)   
84673        4            Waiting to Exhale (1995)   
87429        5  Father of the Bride Part II (1995)   

                                            genres  userId  rating  \
0      Adventure|Animation|Children|Comedy|Fantasy       3     4.0   
49695                   Adventure|Children|Fantasy       1     3.5   
71938                               Comedy|Romance       2     4.0   
84673                         Comedy|Drama|Romance      41     2.0   
87429                                       Comedy      12     2.0   

                 timestamp  index  
0      1999-12-11 13:36:47      1  
49695  2005-04-02 23:53:47  49696  
71938  2000-11-21 15:34:49  71939  
84673  2000-02-27 23:29:58  84674  
87429  1997-03-22 20:49:34  87430

In [10]:
tage = genome_score.merge(genome_tag,on ="tagId")
tage

movieId  tagId  relevance      tag
0               1      1    0.02500      007
1               2      1    0.03975      007
2               3      1    0.04350      007
3               4      1    0.03725      007
4               5      1    0.04200      007
...           ...    ...        ...      ...
11709763   130578   1128    0.01325  zombies
11709764   130840   1128    0.13375  zombies
11709765   131013   1128    0.01625  zombies
11709766   131168   1128    0.01725  zombies
11709767   131170   1128    0.02175  zombies

[11709768 rows x 4 columns]

In [22]:
train_ratings = ratings[['userId', 'movieId', 'rating','title','genres','index']]
train_ratings

Content Based Filtering


In content based filtering we compute the similarity score of users from his past behaviour and recommend new movies based on his previous likings.We can calculate similarity score by using cosine similarity or euclidean similarity and then recommend top 4-5 movies for him.

In this recommender system the content of the movie (overview, cast, crew, keyword, tags,genres etc) is used to find its similarity with other movies. Then the movies that are most likely to be similar are recommended.




Machines cannot understand characters and words. So when dealing with text data we need to represent it in numbers to be

understood by the machine. Countvectorizer is a method to convert text to numerical data.

If I  have 2 text inputs, what happens is that each input is preprocessed, tokenized, and represented as a sparse matrix. By

default, Countvectorizer converts the text to lowercase and uses word-level tokenization.

In [15]:
# Import CountVectorizer and create the count_matrix
from sklearn.feature_extraction.text import CountVectorizer

count = CountVectorizer(stop_words='english')

#Stopwords are the words in any language which does not add much meaning to a sentence.
#They can safely be ignored without sacrificing the meaning of the sentence

count_matrix = count.fit_transform(train_ratings['genres'])

#Visualizing count matrix using pandas dataframe

count_array = count_matrix.toarray()

df = pd.DataFrame(data=count_array,columns = count.get_feature_names())

df

action  adventure  animation  children  comedy  crime  documentary  \
0           0          1          1         1       1      0            0   
1           0          1          0         1       0      0            0   
2           0          0          0         0       1      0            0   
3           0          0          0         0       1      0            0   
4           0          0          0         0       1      0            0   
...       ...        ...        ...       ...     ...    ...          ...   
26739       0          0          0         0       1      0            0   
26740       0          0          0         0       1      0            0   
26741       0          1          0         0       0      0            0   
26742       0          0          0         0       0      0            0   
26743       0          1          0         0       0      0            0   

       drama  fantasy  fi  ...  imax  listed  musical  mystery  noir  romance  \
0          0        1   0  ...     0       0        0        0     0        0   
1          0        1   0  ...     0       0        0        0     0        0   
2          0        0   0  ...     0       0        0        0     0        1   
3          1        0   0  ...     0       0        0        0     0        1   
4          0        0   0  ...     0       0        0        0     0        0   
...      ...      ...  ..  ...   ...     ...      ...      ...   ...      ...   
26739      0        0   0  ...     0       0        0        0     0        0   
26740      0        0   0  ...     0       0        0        0     0        0   
26741      0        0   0  ...     0       0        0        0     0        0   
26742      0        0   0  ...     0       1        0        0     0        0   
26743      0        1   0  ...     0       0        0        0     0        0   

       sci  thriller  war  western  
0        0         0    0        0  
1        0         0    0        0  
2        0         0    0        0  
3        0         0    0        0  
4        0         0    0        0  
...    ...       ...  ...      ...  
26739    0         0    0        0  
26740    0         0    0        0  
26741    0         0    0        0  
26742    0         0    0        0  
26743    0         0    0        0  

[26744 rows x 23 columns]

We can see that 23 words can describe 26744 movies in our dataset. With the count_matrix in our hand we can calculate

similarity score of each movies.Now let us assume we have 26774 vectors describing our movies. Now any movie similar to other

movie can be calculated by using cosine similarity between their vectors.Here, we are taking one movie and calculating

similarity with all movies present in our dataset. Hence our output matrix is of the shape

(np.size(count_matrix),np.size(count_matrix) )  

In [16]:
# Compute the Cosine Similarity matrix based on the count_matrix
from sklearn.metrics.pairwise import cosine_similarity

cosine_sim2 = cosine_similarity(count_matrix, count_matrix)

In [17]:
# Reset index of our main DataFrame and construct reverse mapping
train_ratings = train_ratings.reset_index()
train_ratings

level_0  userId  movieId  rating                               title  \
0             0       3        1     4.0                    Toy Story (1995)   
1         49695       1        2     3.5                      Jumanji (1995)   
2         71938       2        3     4.0             Grumpier Old Men (1995)   
3         84673      41        4     2.0            Waiting to Exhale (1995)   
4         87429      12        5     2.0  Father of the Bride Part II (1995)   
...         ...     ...      ...     ...                                 ...   
26739  20000258   79570   131254     4.0        Kein Bund für's Leben (2007)   
26740  20000259   79570   131256     4.0       Feuer, Eis & Dosenbier (2002)   
26741  20000260   28906   131258     2.5                  The Pirates (2014)   
26742  20000261   65409   131260     3.0                 Rentun Ruusu (2001)   
26743  20000262  133047   131262     4.0                    Innocence (2014)   

                                            genres     index  
0      Adventure|Animation|Children|Comedy|Fantasy         1  
1                       Adventure|Children|Fantasy     49696  
2                                   Comedy|Romance     71939  
3                             Comedy|Drama|Romance     84674  
4                                           Comedy     87430  
...                                            ...       ...  
26739                                       Comedy  20000259  
26740                                       Comedy  20000260  
26741                                    Adventure  20000261  
26742                           (no genres listed)  20000262  
26743                     Adventure|Fantasy|Horror  20000263  

[26744 rows x 7 columns]

In [18]:
indices = pd.Series(train_ratings.index, index=train_ratings['title'])
indices.to_csv('movieslist.csv')
indices

title
Toy Story (1995)                          0
Jumanji (1995)                            1
Grumpier Old Men (1995)                   2
Waiting to Exhale (1995)                  3
Father of the Bride Part II (1995)        4
                                      ...  
Kein Bund für's Leben (2007)          26739
Feuer, Eis & Dosenbier (2002)         26740
The Pirates (2014)                    26741
Rentun Ruusu (2001)                   26742
Innocence (2014)                      26743
Length: 26744, dtype: int64

In [19]:
# Function that takes in movie title as input and outputs most similar movies
def get_recommendations(title, cosine_sim=cosine_sim2):
    # Get the index of the movie that matches the title
    idx = indices[title]
    #print(idx)
    # Get the pairwsie similarity scores of all movies with that movie
    sim_scores = list(enumerate(cosine_sim[idx]))
    #for i in sim_scores:
        #print(i)
    # Sort the movies based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar movies
    sim_scores = sim_scores[1:11]
    #Slicing from one so that the same movie with itself having sim score is not showed
    # Get the movie indices
    movie_indices = [i[0] for i in sim_scores]

    # Return the top 10 most similar movies
    return train_ratings['title'].iloc[movie_indices]

In [20]:
get_recommendations('The Dark Knight (2011)', cosine_sim2)

290     Léon: The Professional (a.k.a. The Professiona...
345                       Clear and Present Danger (1994)
977                              Last Man Standing (1996)
1192            Killer, The (Die xue shuang xiong) (1989)
1522                                      Face/Off (1997)
1537                                      Cop Land (1997)
2455                                Corruptor, The (1999)
2795                               Double Jeopardy (1999)
2873                                    Fight Club (1999)
3178                Hard-Boiled (Lat sau san taam) (1992)
Name: title, dtype: object

In [21]:
get_recommendations('Toy Story (1995)', cosine_sim2)

2209                                           Antz (1998)
3027                                    Toy Story 2 (1999)
3663        Adventures of Rocky and Bullwinkle, The (2000)
3922                      Emperor's New Groove, The (2000)
4790                                 Monsters, Inc. (2001)
10106    DuckTales: The Movie - Treasure of the Lost La...
10978                                     Wild, The (2006)
11861                               Shrek the Third (2007)
13325                       Tale of Despereaux, The (2008)
18230    Asterix and the Vikings (Astérix et les Viking...
Name: title, dtype: object

DRAWBACKS OF THIS RECOMMENDER SYSTEM:

It is only capable of suggesting movies belonging to some genre which the user is searching. It is not able to produce recommendations from different genres capturing user's tastes.

The engine which I built is not so personal, anyone searching for a particular movie would see the same result.So, lets explore other recommender system Collaborative Filtering and get better recommendations than this.